// decentraland-builder-scripts/lib/channel.js
"use strict";
/// <reference path="./types.d.ts" />
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var REQUEST_VALUE = '__request_value__';
var REPLY_VALUE = '__reply_value__';
var POLL_INTERVAL = 5000;
function createChannel(id, host, bus) {
    var handlers = {};
    var requests = {};
    var responses = {};
    bus.on(host.name, function (action) {
        var handler = handlers[action.actionId];
        if (handler) {
            handler(action);
        }
        // clear all pending requests for this entity
        requests = {};
    });
    bus.on(REQUEST_VALUE, function (message) {
        if (message.sender !== id && message.entityName === host.name) {
            var key = message.key;
            var response = responses[key];
            if (response) {
                var value = response();
                var reply = {
                    entityName: host.name,
                    key: key,
                    sender: id,
                    value: value
                };
                bus.emit(REPLY_VALUE, reply);
            }
            // clear pending request for this key
            delete requests[key];
        }
    });
    bus.on(REPLY_VALUE, function (message) {
        if (message.sender !== id && message.entityName === host.name) {
            var key = message.key, value = message.value;
            var request = requests[key];
            if (request) {
                request(value);
            }
            // clear pending request for this key
            delete requests[key];
        }
    });
    return {
        id: id,
        bus: bus,
        createAction: function (actionId, values) {
            var action = {
                entityName: host.name,
                actionId: actionId,
                values: values
            };
            return action;
        },
        sendActions: function (actions) {
            var e_1, _a;
            if (actions === void 0) { actions = []; }
            try {
                for (var actions_1 = __values(actions), actions_1_1 = actions_1.next(); !actions_1_1.done; actions_1_1 = actions_1.next()) {
                    var base = actions_1_1.value;
                    var action = __assign({}, base, { sender: id });
                    bus.emit(action.entityName, action);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (actions_1_1 && !actions_1_1.done && (_a = actions_1.return)) _a.call(actions_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
        },
        handleAction: function (actionId, handler) {
            handlers[actionId] = handler;
        },
        request: function (key, callback) {
            requests[key] = callback;
            var request = { entityName: host.name, key: key, sender: id };
            var interval = setInterval(function () {
                if (key in requests) {
                    bus.emit(REQUEST_VALUE, request);
                }
                else {
                    clearInterval(interval);
                }
            }, POLL_INTERVAL);
        },
        reply: function (key, callback) {
            responses[key] = callback;
        }
    };
}
exports.createChannel = createChannel;
